pyteal/ast/router.py

```diff
@@ -0,0 +1,583 @@
+    ) -> "OnCompleteAction":
+        return OnCompleteAction(on_create=f, on_call=f)
+
+    def is_empty(self) -> bool:
+        return not (self.on_call or self.on_create)
+
+
+OnCompleteAction.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class OnCompleteActions:
+    """
+    OnCompletion Actions keep track of bare-call registrations to all OnCompletion cases.
+    """
+
+    close_out: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    clear_state: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+    delete_application: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+    no_op: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    opt_in: OnCompleteAction = field(kw_only=True, default=OnCompleteAction.never())
+    update_application: OnCompleteAction = field(
+        kw_only=True, default=OnCompleteAction.never()
+    )
+
+    def _clear_state_n_other_oc(
+        self,
+    ) -> tuple[OnCompleteAction, dict[EnumInt, OnCompleteAction]]:
+        return self.clear_state, {
+            OnComplete.CloseOut: self.close_out,
+            OnComplete.DeleteApplication: self.delete_application,
+            OnComplete.NoOp: self.no_op,
+            OnComplete.OptIn: self.opt_in,
+            OnComplete.UpdateApplication: self.update_application,
+        }
+
+    def is_empty(self) -> bool:
+        for action_field in fields(self):
+            action: OnCompleteAction = getattr(self, action_field.name)
+            if not action.is_empty():
+                return False
+        return True
+
+
+OnCompleteActions.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class CondNode:
+    condition: Expr
+    branch: Expr
+
+
+CondNode.__module__ = "pyteal"
+
+
+@dataclass
+class CategorizedCondNodes:
+    method_calls_create: list[CondNode] = field(default_factory=list)
+    bare_calls_create: list[CondNode] = field(default_factory=list)
+    method_calls: list[CondNode] = field(default_factory=list)
+    bare_calls: list[CondNode] = field(default_factory=list)
+
+    def program_construction(self) -> Expr:
+        concatenated_ast = (
+            self.method_calls_create
+            + self.bare_calls_create
+            + self.method_calls
+            + self.bare_calls
+        )
+        if not concatenated_ast:
+            raise TealInputError("ABIRouter: Cannot build program with an empty AST")
+        program: Cond = Cond(*[[n.condition, n.branch] for n in concatenated_ast])
+        return program
+
+
+CategorizedCondNodes.__module__ = "pyteal"
+
+
+class Router:
+    """
+    Class that help constructs:
+    - a *Generalized* ARC-4 app's approval/clear-state programs
+    - and a contract JSON object allowing for easily read and call methods in the contract
+    """
+
+    def __init__(
+        self,
+        name: str,
+        bare_calls: OnCompleteActions,
```

Comment:
Support for bare app calls are optional for contracts, so I think this object should be optional

Reply:
👍